rust/agama-network/src/model.rs

```diff
@@ -0,0 +1,439 @@
+            uuid,
+            ..Default::default()
+        };
+        let conn2 = Connection::Ethernet(EthernetConnection { base: base1 });
+        state.update_connection(conn2).unwrap();
+        let found = state.get_connection(uuid).unwrap();
+        assert_eq!(found.uuid(), uuid);
+    }
+
+    #[test]
+    fn test_update_unknown_connection() {
+        let mut state = NetworkState::default();
+        let uuid = Uuid::new_v4();
+        let base = BaseConnection {
+            uuid,
+            ..Default::default()
+        };
+        let conn0 = Connection::Ethernet(EthernetConnection { base });
+        let error = state.update_connection(conn0).unwrap_err();
+        assert!(matches!(error, NetworkStateError::UnknownConnection(_)));
+    }
+
+    #[test]
+    fn test_remove_connection() {
+        let mut state = NetworkState::default();
+        let uuid = Uuid::new_v4();
+        let base0 = BaseConnection {
+            uuid,
+            ..Default::default()
+        };
+        let conn0 = Connection::Ethernet(EthernetConnection { base: base0 });
+        state.add_connection(conn0).unwrap();
+        state.remove_connection(uuid).unwrap();
+        let found = state.get_connection(uuid).unwrap();
+        assert!(found.is_missing());
+    }
+
+    #[test]
+    fn test_remove_unknown_connection() {
+        let mut state = NetworkState::default();
+        let error = state.remove_connection(Uuid::new_v4()).unwrap_err();
+        assert!(matches!(error, NetworkStateError::UnknownConnection(_)));
+    }
+}
+
+/// Network device
+#[derive(Debug, Clone)]
+pub struct Device {
+    pub name: String,
+    pub ty: DeviceType,
+}
+
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum DeviceType {
+    Loopback = 0,
+    Ethernet = 1,
+    Wireless = 2,
+    Unknown = 3,
+}
+
+impl TryFrom<u8> for DeviceType {
+    type Error = NetworkStateError;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(DeviceType::Loopback),
+            1 => Ok(DeviceType::Ethernet),
+            2 => Ok(DeviceType::Wireless),
+            3 => Ok(DeviceType::Unknown),
+            _ => Err(NetworkStateError::InvalidDeviceType(value)),
+        }
+    }
+}
+
+/// Represents an available network connection
+#[derive(Debug, PartialEq, Clone)]
+pub enum Connection {
+    Ethernet(EthernetConnection),
+    Wireless(WirelessConnection),
+    Loopback(LoopbackConnection),
+}
+
+impl Connection {
+    pub fn new(id: String, device_type: DeviceType) -> Self {
+        let base = BaseConnection {
+            id: id.to_string(),
+            ..Default::default()
+        };
+        match device_type {
+            DeviceType::Wireless => Connection::Wireless(WirelessConnection {
+                base,
+                ..Default::default()
+            }),
+            _ => Connection::Ethernet(EthernetConnection { base }),
+        }
+    }
+
+    pub fn base(&self) -> &BaseConnection {
+        match &self {
```

Comment:
is it needed? arg there easier way to have method in this kind of inheritance? Or it needs macros or traits to do it?

Reply:
Sure, I already thought about it. But I prefer not to add more complexity to this PR (adding a macro and so on). I will add a TODO so we can implement that in the future. It might be a nice exercise 😃

Reply:
yeap. I think when we start adding more connection types, it will be needed to make it as easy as possible without touching too man placing. TODO is fine for me.